proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,725 @@
+index spaces.
+
+Other than type/module sharing, which can easily be undone by duplication, a
+module A that imports module B is semantically equivalent to module A nesting
+module B. This means tools can merge multiple `.wasm` files into one `.wasm` by
+a trivial, language-independent transformation. Thus, module imports and nested
+modules can serve as a language-independent basis for first expressing a module
+dependency graph and then bundling/linking that dependency graph into a single
+`.wasm` module for portable distribution.
+
+Additionally, on the Web and other hosts that provide runtime instantiation
+APIs, nested modules allow an outer module to bundle module code that is to be
+instantiated dynamically by passing a first-class module reference (created via
+`ref.module`) to the runtime instantiation API.
+
+
+### Module and Instance Exports
+
+Lastly, symmetric to all other kinds of definitions, modules and instances can
+be exported. For example:
+```wasm
+(module
+  (import "a" (module $a ...))
+  (module $b ...)
+  (import "c" (instance $c ...))
+  (instance $d ...)
+
+  (export "e1" (module $a))
+  (export "e2" (module $b))
+  (export "e3" (instance $c))
+  (export "e4" (instance $d))
+)
+```
+Therefore, module and instance types can appear in both the imports and exports
+of module types and instance types.
+
+
+### Summary
+
+To summarize the proposed changes (all changes in both text and binary format):
+* The `module` field of [`import`] become optional (allowing single-level
+  imports). (How to encode this in the [import section] is an interesting
+  question.)
+* New `module` and `instance` type constructors are added that may be used to
+  define types in the [type section].
+* A new `alias` case is added to type definitions allowing the import of parent
+  modules' type definitions.
+* New `module` and `instance` cases are added to [`importdesc`], with the
+  type indicated by `typeidx`, like `func`.
+* A new `module` section is added which contains a sequence of either module
+  definitions or `alias` definitions (of parents' modules). In terms of the
+  binary format, a nested module definition has the same binary format as a
+  non-nested `.wasm` module. Thus, the [`module`] binary format production
+  becomes recursive.
+* A new `instance` section is added which contains a sequence of instance
+  definitions. Like [`global`] definitions, an `instance` definition contains
+  an `instance` type and an `init` expression which is validated to match.
+* A new `instance.instantiate` instruction is added which is only allowed in
+  `instance` `init` expressions.
+* New `ref.instance` and `ref.module` instructions are added to the general
+  instruction set.
+* New module and instance reference types `(ref (module ...))`, `moduleref`,
+  `(ref (instance ..))` and `instanceref` are added to [`reftype`].
+* New `module` and `instance` cases are added to [`exportdesc`].
```

Comment:
Should this list `ref.memory` and friends?

Reply:
Good catch, done.